File: 2.add-two-numbers.ts
/*
 * @lc app=leetcode id=2 lang=typescript
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const head = new ListNode();
    var currentNode = head;
    var currentNodel1 = l1;
    var currentNodel2 = l2;
    var carry = 0;

    while (true) {
        const digit1 = currentNodel1?.val || 0;
        const digit2 = currentNodel2?.val || 0;
        currentNode.val = (digit1 + digit2 + carry) % 10;
        carry = Math.floor((digit1 + digit2 + carry) / 10);
        if (carry === 0 && (!currentNodel1 || !currentNodel1?.next) && (!currentNodel2 || !currentNodel2?.next)) break;
        currentNode.next = new ListNode();
        currentNode = currentNode.next;
        currentNodel1 = currentNodel1 ? currentNodel1.next : null;
        currentNodel2 = currentNodel2 ? currentNodel2.next : null;
    }

    return head;
};
// @lc code=end

File: 2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        node1, node2 = l1, l2
        dummy_result_head = ListNode()
        result_node = dummy_result_head
        while node1 or node2:
            value = 0
            if node1:
                value += node1.val
                node1 = node1.next
            if node2:
                value += node2.val
                node2 = node2.next
            value += carry
            carry = value // 10
            value = value % 10
            result = ListNode(value)
            result_node.next = result
            result_node = result_node.next
        if carry > 0:
            result_node.next = ListNode(carry)

        return dummy_result_head.next

# @lc code=end

